Correctly resolve dispatch keys for PyOperator

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96306
* #96304
* #96302

Previously, we never actually used resolve_key, which meant that
you had to register CPU/CUDA/etc all manually; none of the alias
keys worked.  Now they work.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Correctly resolve dispatch keys for PyOperator

Previously, we never actually used resolve_key, which meant that
you had to register CPU/CUDA/etc all manually; none of the alias
keys worked.  Now they work.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Correctly resolve dispatch keys for PyOperator"


Previously, we never actually used resolve_key, which meant that
you had to register CPU/CUDA/etc all manually; none of the alias
keys worked.  Now they work.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Correctly resolve dispatch keys for PyOperator"


Previously, we never actually used resolve_key, which meant that
you had to register CPU/CUDA/etc all manually; none of the alias
keys worked.  Now they work.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

zou3519 commented on torch/_ops.py:
```diff
@@ -218,8 +222,14 @@ def dispatch(self, dispatch_key, *args, **kwargs):
-        return self.py_kernels[dispatch_key](*args, **kwargs)
+        final_key = resolve_key(self, dispatch_key)
+
+        assert not isinstance(
+            final_key, str
+        ), f"C++ fallbacks don't work for PyOperator, please reimplement {final_key}"
```
I might not be reading this right, but how can `resolve_key` return a str? It seems to only return torch._C.DispatchKey

ezyang replied:
oops, you're right

zou3519 reviewed (approved):
LGTM

ezyang committed:
Update on "Correctly resolve dispatch keys for PyOperator"


Previously, we never actually used resolve_key, which meant that
you had to register CPU/CUDA/etc all manually; none of the alias
keys worked.  Now they work.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Correctly resolve dispatch keys for PyOperator"


Previously, we never actually used resolve_key, which meant that
you had to register CPU/CUDA/etc all manually; none of the alias
keys worked.  Now they work.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Correctly resolve dispatch keys for PyOperator"


Previously, we never actually used resolve_key, which meant that
you had to register CPU/CUDA/etc all manually; none of the alias
keys worked.  Now they work.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
